### PR TITLE
feat: continuation of update @dimforge/rapier3d-compat to 0.12.0 + optimizations

### PR DESCRIPTION
**Pull request by 0xtito**
## Description
- Continuation of @isaac-mason's draft 
  - #601 
- Added examples for new `spring` and `rope` joints
  - Spring Joint example (based on example from [rapier](https://github.com/dimforge/rapier/pull/579))
 
https://github.com/pmndrs/react-three-rapier/assets/92818759/945157bb-24bd-4024-b4aa-c5d2c0ec4c0e

  - Rope Joint example

https://github.com/pmndrs/react-three-rapier/assets/92818759/47725b4b-d216-43e7-a710-f91a89494f2d

- Changed the type for `body1Anchor` and `body2Anchor` to react-three/fiber's `Vector3` + created `vector3ToRapierVector`.
  - I did not make changes to all joint params, only for the `spring` and `rope` joints.
  - This was done because I believe we are unnecessarily creating threejs `Vector3` objects, when all that is required is a `Vector3Tuple`, but a threejs Vector3 also works. So now you can pass either a threejs `Vector3`, a `Vector3Tuple`, or a `number`. If it is a `Vector3Tuple`, then we are just creating a `RapierVector` which is essentially an object with `x`, `y`, and `z` as props.
- Added setter for `additionalSolverIterations` to `mutableRigidBodyOptions`
- Change default values for `erp` and `predictionDistance` to `0.8` and `0.002`, respectfully.
  - to be honest, I completely understand why we changed to default value of `erp` to `0.2`. Inside of rapier.js, the default value is `0.2`, as shown [here](https://github.com/dimforge/rapier.js/blob/98015b1871f4da6ba320f9ec6f34fe34367b088b/src.ts/dynamics/integration_parameters.ts#L31).
  

### Changes
- Update `<Physics />` to reflect to `integrationParameters` for `@dimforge/rapier3d-compat` v0.12.0
- Updated `useSpringJoint` and `useRopeJoint` started by @isaac-mason
  - Will need to add some good docs/comments for these because the order of the passed in `bodyAnchor`s is significant
- Enhanced `<SpringExample />` and `<RopeJointExample />` examples
- Created `vector3ToRapierVector` utility function
- Testing changing `bodyAnchor` types to react-three/fiber's `Vector3`.
  - This will allow someone to pass either a threejs `Vector3`, `Vector3Tuple`, or a `number`.
  - if there is no errors/bugs, we should change all JointParams to react-three/fiber's `Vector3` so we do not unnecessarily create threejs `Vector3` classes.

### Checklist:
- [x] 🔍 I have performed a self-review of my code
- [ ] 💬 I have commented my code, particularly in hard-to-understand areas
- [ ] 📗 I have made corresponding changes to the documentation
- [x] ⭐️ My changes generate no new warnings
- [ ] 🧪 I have added tests
- [ ] 🟢 All new and existing unit tests pass

### CODE REVIEW

File: packages/react-three-rapier/src/components/Physics.tsx
Comment (isaac-mason) [inline packages/react-three-rapier/src/components/Physics.tsx:319]: Seems the actual erp default might be 0.6?

https://github.com/dimforge/rapier/blob/da92e5c2837b27433286cf0dd9d887fd44dda254/src/dynamics/integration_parameters.rs#L197

Maybe the rapier.js jsdoc is just out of date

File: packages/react-three-rapier/src/utils/utils.ts
Comment (isaac-mason) [inline packages/react-three-rapier/src/utils/utils.ts:104]: imo we should change all joint hooks to be consistent before releasing so the API surface is consistent. We can do that in a separate PR though.

**Comment by isaac-mason** [comment]
Before releasing we'll also want to docs for the new joints to README.md

**Comment by isaac-mason** [comment]
Also realised we can add the "Generic" joint too - but again, doesn't have to all be in this PR

> Added JointData.generic (3D only) to create a generic 6-dof joint and manually specify the locked axes.

https://github.com/dimforge/rapier.js/blob/master/CHANGELOG.md

**Comment by isaac-mason** [comment]
@0xtito added some docs to readme.md now 🙂

File: packages/react-three-rapier/src/utils/utils.ts
Comment (isaac-mason) [inline packages/react-three-rapier/src/utils/utils.ts:110]: It's probably best to always construct a rapier Vector3 for interacting with the rapier api.
While this may work now (haven't tested), rapier could make changes in future that break this.

File: packages/react-three-rapier/src/components/Physics.tsx
Comment (isaac-mason) [reply packages/react-three-rapier/src/components/Physics.tsx:319]: I think leaving this default untouched for this release is ok though

**Comment by isaac-mason** [comment]
Let me know if the changes I added look good to you @0xtito 🙂

**Comment by drcmda** [comment]
beautiful! thank you all!

File: packages/react-three-rapier/src/components/Physics.tsx
Comment (0xtito) [reply packages/react-three-rapier/src/components/Physics.tsx:319]: No stress! I should have checked the source integration_paramaters file earlier.

And I agree, `0.8` is good for now.

File: packages/react-three-rapier/src/utils/utils.ts
Comment (0xtito) [reply packages/react-three-rapier/src/utils/utils.ts]: Okay nice thank you ! I was hesitant to extend the changes to all joint hooks until I knew the changes were sound.

File: packages/react-three-rapier/src/utils/utils.ts
Comment (0xtito) [reply packages/react-three-rapier/src/utils/utils.ts]: Ya good call - it also did look odd to the return either a THREE.Vector3 or Rapier Vector. This is likely the right approach. 👍

**Comment by 0xtito** [comment]
> Let me know if the changes I added look good to you @0xtito 🙂

Looks great to me! Thanks for the additions.

There is one thing that is kinda bothering me though. The entire simulation just feels slower compared to the rapier/rapier.js examples. For example in the wrecking ball video above, the ball seems to move slow, and even when the boxes are hit they seem to fall slowly.

This isn't necessarily a problem, but I do wonder why this is the case. I also think this could be discussed/dissected in another PR.

**Comment by 0xtito** [comment]
just changed the text for `<RopeJointExample />` button to "Rope Joint" just to keep the styling in the demo consistent.

**Comment by 0xtito** [comment]
is there anything else that needs to be added before the PR can be merged? I left the checklist as is, but does the default checklist apply here?

**Comment by 0xtito** [comment]
Marked as ready for review - if there are still things needed to be done feel free to change it back to a draft!

**Comment by isaac-mason** [comment]
These changes look good to me! Let's get a review from @wiledal too.

> There is one thing that is kinda bothering me though. The entire simulation just feels slower compared to the rapier/rapier.js examples. For example in the wrecking ball video above, the ball seems to move slow, and even when the boxes are hit they seem to fall slowly.

I can take a look at this shortly.

**Comment by wiledal** [comment]
Changes look good to me! I'm wondering though why the performance has degraded and is now noticeably stuttery on my machine compared to the old demo -- new settings defaults? (possibly due to the `erp` value change)

Also, the contact force events are not firing like they used to 🤔 
- current: https://react-three-rapier.pmnd.rs/contact-force-events
- new: https://react-three-rapier-md77pr5an-pmndrs.vercel.app/contact-force-events

**Comment by 0xtito** [comment]
> Changes look good to me! I'm wondering though why the performance has degraded and is now noticeably stuttery on my machine compared to the old demo -- new settings defaults? (possibly due to the `erp` value change)
> 
> Also, the contact force events are not firing like they used to 🤔
> 
> * current: https://react-three-rapier.pmnd.rs/contact-force-events
> * new: https://react-three-rapier-md77pr5an-pmndrs.vercel.app/contact-force-events

Regarding the performance, the performance seems to be the exact same in both environments for me. Do you think it could it could be a browser problem? I am using Arc, but could test it across different browsers to see. Not sure why that would be happening.

Also regarding the contact force, that is odd. In the new `contact-force-events` demo, it only emits the contact force once the ball is starts to bounce very little, and only then does it bounce higher + emit the contact force to the console. I'll try to debug this now.

**Comment by 0xtito** [comment]
I also realized the Spring and Joint sections were missing from the Readme topics so I just added them

**Comment by wiledal** [comment]
I'm on Arc as well. I double checked, and it does not stutter anymore. Sometimes I seem to get a stuttering version after random refreshes, that might not have anything to do with this update however.

I'm looking at the contact force event issue. It seems it just doesn't fire anymore unless the body is almost at rest. Unclear why that would be.

**Comment by wiledal** [comment]
There's a clear difference between the contact force event from Rapier.js 0.11.2 and 0.12.0.

<table>
<td>
0.11.2

Screenshot 2024-02-07 at 22 12 34

</td>

<td>
0.12.0

Screenshot 2024-02-07 at 22 13 56

</td>
</table>

This is a plain vite setup with `rapier3d-compat`, and nothing else.
It's odd that we are hardly getting the collision events at all in r3/rapier tho... I'll investigate a bit more, but we might continue without this for now.

**Comment by 0xtito** [comment]
Mhmmm thats so weird. 

I noticed in rapier/rapier.js there is a `contactForceEventThreshold` get/set on the collider class. I do not think we have that implemented..? Assuming its not already implemented, I am going to do so and try playing with the value to see if the problem is that the threshold is not being met

**Comment by 0xtito** [comment]
Just mentioning this change, I am unsure if this is directly related, but in rapier.js v12, the names of `contactsWith` and `intersectionsWith` have been changed to `contactPairsWith` and `intersectionPairsWith`, respectively.
image

link to PR [here](https://github.com/dimforge/rapier.js/pull/252/files#diff-04781526dfce9f2f63f1c981f7c6694dcbb2c649d147bc850f179d9de01cd725)

**Comment by 0xtito** [comment]
Another thing I have noticed is that, in the new r3-rapier, if you just let the `ContactForceEvents` example run for a while and watch the log - the contactForce behaves very erratically. Like it will seemingly randomly fire off the `onContactForce` event. Sometimes launching the ball in back to back bounces.

So I changed the default `numSolverIterations` to `1` in the `Physics.tsx` file to see if it was a problem with the new solver/solverIterations. And that did the trick, now every bounce the `onContactForce` callback is running. On top of that, the `totalForceMagnitude` has the values we are used to (the higher values).

I am going to keep dissecting this to see why the `numSolverIterations` is exactly causing this.

**Comment by isaac-mason** [comment]
Let's also see if anyone on the rapier discord can help us, posted a message now: https://discord.com/channels/507548572338880513/748814906953826334/1204933879677329488

**Comment by isaac-mason** [comment]
Also created an issue: https://github.com/dimforge/rapier.js/issues/261

**Review by wiledal — APPROVED** [review]

**Comment by 0xtito** [comment]
Sounds great! Ya it definitely seems to be a problem with Rapier. Once they address the issue, we can merge in those changes then.

Also, there are a few other things we need to add to fully upgrade to 0.12 (like integrating the `switchToStandardPgsSolver` and `switchToSmallStepsPgsSolver` methods, adding the Generic JointData, add `DynamicRayCastVehicleController`, and  update the `CharacterController` with the changes specified [here](https://github.com/dimforge/rapier.js/pull/252)), but what we have now seems more than enough to merge this PR!

**Comment by 0xtito** [comment]
Let me know @wiledal if there is anything else I can add/do for this PR 🙂

**Comment by 0xtito** [comment]
Hey @isaac-mason, do you think this PR is ready to be merged? No rush if you would like to make some additions/changes, but if not I would love for this to be merged so I can start testing it out in my library 🙂

**Comment by wiledal** [comment]
I will put the repo in `canary` mode, and merge this so that it can be tested without breaking!

**Comment by 0xtito** [comment]
Great! Thanks @wiledal - I am using the spring joint to try to build a rapier-integrated hand for mixed reality so i will definitely be testing some edge cases. If I find any errors I will mention them here or as an issue, dependent on the error/your preference (sorry for any hassle i've caused, this is my first time meaningfully contributing to OS and I don't know the right course of action sometimes lol)

**Comment by wiledal** [comment]
I've done some initial tests in `canary`, and I think it's seems stable to release normally! Thanks for the awesome work, and for keeping at it. 

You should join us on the Poimandres discord for easier communication outside of the pull request context 😄  https://discord.com/channels/740090768164651008/969941082144124959

**Pull request merged**
